File: internal/app/dispatch.go
package app

import (
	"github.com/leandronsp/githerb/internal/review"
)

// Dispatch hands the open notes on a proposal to an agent. It writes the
// request down rather than starting anything, because whoever runs the agent
// is a separate process that may not be up yet.
type Dispatch struct {
	Proposals review.Proposals
	Author    string
	Now       Clock
}

// Run records the request against the head revision.
func (d Dispatch) Run(id string) (review.Proposal, error) {
	proposal, err := d.Proposals.Load(review.ProposalID(id))
	if err != nil {
		return review.Proposal{}, err
	}

	head := proposal.Head().SHA()

	ask, err := review.NewDispatch(head, d.Author, d.Now())
	if err != nil {
		return review.Proposal{}, err
	}

	if err := d.Proposals.Annotate(head, review.DispatchRecord(ask)); err != nil {
		return review.Proposal{}, err
	}

	return proposal, nil
}

File: crates/app/src/dispatch.rs
//! Handing the open notes to an agent: a person saying "somebody answer
//! these", which the runner reads as a job.

use review::{Author, Dispatch, Proposal, ProposalId, Record, Timestamp};

use crate::error::Result;
use crate::store::Store;

/// Ask for an agent on the head revision.
///
/// # Errors
///
/// A proposal nobody opened, or one that is no longer open.
pub fn dispatch(
    store: &Store,
    author: &Author,
    now: Timestamp,
    id: &ProposalId,
) -> Result<Proposal> {
    let mut proposal = store.load(id)?;
    let head = proposal.head().sha().clone();
    let record = Record::Dispatch(Dispatch::new(head.clone(), author.clone(), now));

    proposal.apply(record.clone())?;
    store.annotate(&head, &record)?;

    Ok(proposal)
}
